Garden: Domain mapping plan check

markbiek opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of ARC-1121
Part of ARC-1093

## Proposed Changes

In https://github.com/Automattic/wp-calypso/pull/106489, we made a change so that we skipped checkout for mapping a domain for all garden sites (as a temporary workaround). 

Now we have 194871-ghe-Automattic/wpcom which adds domain mapping features for the appropriate CIAB plans so we're changing the `/ciab` dashboard to do a plan check to decide if we can skip checkout.

Garden sites will be redirected back to `/ciab/sites/somesite.com/domains` while other sites will be redirected to the normal domain management page.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Create a new garden site using the `/tools/garden` MC tool.
* Add a `woo_hosted_basic_monthly` plan to the site by going to `https://wordpress.com/checkout/your-new-site.commerce-garden.com/woo_hosted_basic_plan_monthly/`.
* Now use the calypso.live url (or run this locally) and go to `/ciab/sites/your-new-site.commerce-garden.com/domains`.
* You should see the following banner instead of the banner saying that changing the primary site address requires a plan.
<img width="754" height="190" alt="CleanShot 2025-10-29 at 14 14 04" src="https://github.com/user-attachments/assets/9d51bb87-bff5-4ad8-8726-98d2d492eef2" />

* Click "Add domain name" and then "Use a domain name I own", then enter a domain name. (I've been using `random-subdomain.biek.dev` but it doesn't really matter since we're not configuring DNS for the domain).
* Click "Continue", then click "Select" under "Connect your domain".
* Your should return to the site overview.
* Click the "Domains" link at the top of the page.
* Click on the "..." Actions menu for the newly mapped domain.
* Click "Make primary site address".
  - Note that there's a 30 minute delay before you can set a new domain to primary. You can skip this by modifying [this line](https://github.com/Automattic/wp-calypso/blob/trunk/client/dashboard/domains/dataviews/actions.tsx#L194) from `! isRecentlyRegistered( item.registration_date )` to `( true || ! isRecentlyRegistered( item.registration_date ) )`. You may also need to comment out line 1244 of `class.wpcom-store-domains-api-endpoints.php` on your sandbox and sandbox the public-api.
* The domain should be made primary

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

markbiek committed:
Revert "Allow garden sites to a map a domain from /ciab"

This reverts commit 5aaa8d12f308435de5451c195e8c4104c7fe1d91.

We'll be changing from a garden-specific check to a plan feature check.

WPprodigy commented on client/landing/stepper/declarative-flow/flows/domain/domain.ts:
```diff
@@ -133,23 +134,30 @@ const domain: FlowV2< typeof initialize > = {
-						) {
+						const isDomainMapping =
+							providedDependencies.domainCartItem?.product_slug === 'domain_map';
+						const mappingIsFree = isDomainMappingFree( site );
```
This [isDomainMappingFree()](https://github.com/Automattic/wp-calypso/blob/2fb3b86694c5d6416232a2da9d0dbf4ff8326081/client/lib/cart-values/cart-items.ts#L838-L844) function is a bit suspicious in it's contents. It's also notably always paired with these other [two](https://github.com/Automattic/wp-calypso/blob/2fb3b86694c5d6416232a2da9d0dbf4ff8326081/client/components/domains/use-my-domain/utilities/get-mapping-free-text.js#L18-L21) [functions](https://github.com/Automattic/wp-calypso/blob/2fb3b86694c5d6416232a2da9d0dbf4ff8326081/client/components/domains/use-my-domain/utilities/get-mapping-price-text.js#L13-L15).

At least for our use case, should we instead just check if the site already has a plan that includes the `DOMAIN_MAPPING` feature (`hasPlanFeature( site, DotcomFeatures.DOMAIN_MAPPING `)? That way we can just either stick the feature on the free/trial site plans or not. The gating will adjust accordingly.

WPprodigy commented on client/landing/stepper/declarative-flow/flows/domain/domain.ts:
```diff
@@ -133,23 +134,30 @@ const domain: FlowV2< typeof initialize > = {
+									? `/ciab/sites/${ siteSlug }/domains`
+									: `/domains/manage/${ siteSlug }`;
+
+								return window.location.assign( redirectUrl );
```
I think we're in need of a more generic way of knowing which dashboard instance we should be sending the user to. There's going to be more than just `/ciab/` (which itself is temporary). Seems like we need to know which HD the user came from so that we can return there? `#multi-site-dashboard-ciab` in slack might be able to help if you're unsure on the right approach.

markbiek replied:
My first thought was using a `redirect_to` query param (which we do elsewhere). Folks in `#multi-site-dashboard-ciab` confirmed that approach.

markbiek committed:
Add domain mapping feature to feature-list

markbiek committed:
Tweak hasPlanFeature to accept either a Site or SiteDetails

markbiek committed:
Update post-domain mapping redirect

For CIAB sites, we map the domain and redirect without checking out if
they already have a plan.

Previously, we had an isGarden check which redirected back to the /ciab
dashboard. Now we use a redirect_to query parameter.

WPprodigy commented on client/dashboard/sites/domains/index.tsx:
```diff
@@ -51,12 +51,14 @@ function SiteDomains() {
 	);
 
+	const redirectTo = `/ciab/sites/${ site.slug }/domains`;
+
```
Can't do this (at least not here) since we're not in a ciab-specific component.

markbiek replied:
Good call. I can change it to look at the url instead.

WPprodigy commented on client/dashboard/utils/site-features.ts:
```diff
@@ -9,7 +9,7 @@ import type {
 // Returns whether the plan supports a specific feature.
 export function hasPlanFeature(
-	site: Site,
+	site: { plan?: { features: { active: string[] } } },
```
Just curious why the signature change

markbiek replied:
I needed `hasPlanFeature` to accept either a `Site` or `SiteDetails` object. The `plan.features` part is common between the two so it made sense to me to just accept an object with the data the function requires.

markbiek committed:
Update SiteDomains redirect to use the current url

WPprodigy reviewed (approved):
Awesome, this is looking good. Is nice and generic now - not even a mention of "garden" :)
